src/main/java/org/opentripplanner/transit/model/timetable/ScheduledTripTimes.java

```diff
@@ -0,0 +1,352 @@
+  public boolean isRecordedStop(int stop) {
+    return false;
+  }
+
+  @Override
+  public boolean isNoDataStop(int stop) {
+    return false;
+  }
+
+  @Override
+  public boolean isPredictionInaccurate(int stop) {
+    return false;
+  }
+
+  @Override
+  public I18NString getTripHeadsign() {
+    return trip.getHeadsign();
+  }
+
+  @Override
+  @Nullable
+  public I18NString getHeadsign(final int stop) {
+    return (headsigns != null && headsigns[stop] != null)
+      ? headsigns[stop]
+      : getTrip().getHeadsign();
+  }
+
+  @Override
+  public List<String> getHeadsignVias(final int stop) {
+    if (headsignVias == null || headsignVias[stop] == null) {
+      return List.of();
+    }
+    return List.of(headsignVias[stop]);
+  }
+
+  @Override
+  public int getNumStops() {
+    return arrivalTimes.length;
+  }
+
+  @Override
+  public Accessibility getWheelchairAccessibility() {
+    return trip.getWheelchairBoarding();
+  }
+
+  @Override
+  public OccupancyStatus getOccupancyStatus(int ignore) {
+    return OccupancyStatus.NO_DATA_AVAILABLE;
+  }
+
+  @Override
+  public int gtfsSequenceOfStopIndex(final int stop) {
+    return originalGtfsStopSequence[stop];
+  }
+
+  @Override
+  public OptionalInt stopIndexOfGtfsSequence(int stopSequence) {
+    if (originalGtfsStopSequence == null) {
+      return OptionalInt.empty();
+    }
+    for (int i = 0; i < originalGtfsStopSequence.length; i++) {
+      var sequence = originalGtfsStopSequence[i];
+      if (sequence == stopSequence) {
+        return OptionalInt.of(i);
+      }
+    }
+    return OptionalInt.empty();
+  }
+
+  @Override
+  public boolean equals(Object o) {
+    throw new UnsupportedOperationException("Not implemented, implement if needed!");
+  }
+
+  @Override
+  public int hashCode() {
+    throw new UnsupportedOperationException("Not implemented, implement if needed!");
+  }
+
+  /* package local - only visible to timetable classes */
+
+  int[] copyArrivalTimes() {
+    return IntUtils.shiftArray(timeShift, arrivalTimes);
+  }
+
+  int[] copyDepartureTimes() {
+    return IntUtils.shiftArray(timeShift, departureTimes);
+  }
+
+  I18NString[] copyHeadsigns(Supplier<I18NString[]> defaultValue) {
+    return headsigns == null ? defaultValue.get() : Arrays.copyOf(headsigns, headsigns.length);
+  }
+
+  /* private methods */
+
+  private void validate() {
+    int lastStop = departureTimes.length - 1;
+    IntUtils.requireInRange(departureTimes[0], MIN_TIME, MAX_TIME);
+    IntUtils.requireInRange(arrivalTimes[lastStop], MIN_TIME, MAX_TIME);
+    validateNonIncreasingTimes();
+  }
+
+  /**
+   * When creating scheduled trip times we could potentially imply negative running or dwell times.
+   * We really don't want those being used in routing. This method checks that all times are
+   * increasing. The first stop arrival time and the last stops depature time is NOT checked -
+   * these should be ignored by raptor.
+   */
+  private void validateNonIncreasingTimes() {
```

Comment:
Looks like this could be a common logic between RealtimeTripTimes and ScheduledTripTimes

Reply:
Yes, I am not done with the refactoring - there is a lot of duplication and I have intentionally not made any changes here yet. I am refactoring the code using TripTimes first, then when I am done I know what code should be deleted, shared and kept in RT and Scheduled TripTimes - so for know ignore the duplication.